src/rust/runtime/memory/buffer/demibuffer.rs

```diff
@@ -0,0 +1,820 @@
+// In particular, len(), adjust(), trim(), clone(), and split() are present and designed to behave the same regardless.
+//
+// The constructors/destructors, however, are necessarily different.  For DPDK-allocated buffers, a MBuf is expected
+// to be allocated externally and provided to the DemiBuffer's "from_mbuf" constructor.  A MBuf can also be extracted
+// from a DPDK allocated DemiBuffer via the "into_mbuf" routine.
+//
+// Note: if compiled without the "libdpdk" feature defined, the DPDK-specific functionality won't be present.
+
+// Note on buffer chain support:
+// DPDK has a concept of MBuf chaining where multiple MBufs may be linked together to form a "packet".  While the
+// DemiBuffer routines for heap-allocated buffers also now support this functionality, it isn't yet exposed via the
+// DemiBuffer interface.
+// ToDo: Expose buffer chain support once we have a solid use case.
+
+// Note on intrusive queueing:
+// Since all DemiBuffer types keep the metadata for each "view" in a separate allocated region, they can be queued
+// using intrusive links (i.e. have a link field in the metadata).
+// ToDo: Expose calls to get/set a linking field.
+
+// Note on the allocation functions:
+// This code currently uses std::alloc() and std::dealloc() to allocate/free things from the heap.  Note that the Rust
+// documentation says that these functions are expected to be deprecated in favor of their respective methods of the
+// "Global" type when it and the "Allocator" trait become stable.
+use crate::runtime::fail::Fail;
+#[cfg(feature = "libdpdk")]
+use ::dpdk_rs::{
+    rte_mbuf,
+    rte_mempool,
+    rte_pktmbuf_adj,
+    rte_pktmbuf_clone,
+    rte_pktmbuf_free,
+    rte_pktmbuf_trim,
+};
+#[cfg(feature = "libdpdk")]
+use ::std::mem;
+use ::std::{
+    alloc::{
+        alloc,
+        dealloc,
+        handle_alloc_error,
+        Layout,
+    },
+    marker::PhantomData,
+    mem::size_of,
+    num::NonZeroUsize,
+    ops::{
+        Deref,
+        DerefMut,
+    },
+    ptr::{
+        self,
+        null_mut,
+        NonNull,
+    },
+    slice,
+};
+
+// Cache line size, defined here for code clarity.
+// While this is arguably architecture dependent, and thus should be pulled from some architecture-specific information,
+// this value (64 bytes) is used by all architectures we're likely to care about (and is also assumed by DPDK).
+const CACHE_LINE_SIZE: usize = 64;
+
+// Buffer Metadata.
+// This is defined to match a DPDK MBuf (rte_mbuf) in order to potentially use the same code for some DemiBuffer
+// operations that currently use identical (but separate) implementations for heap vs DPDK allocated buffers.
+// Fields beginning with an underscore are not directly used by the current DemiBuffer implementation.
+// Should be cache-line aligned and consume 2 cache lines (128 bytes).
+#[repr(C)]
+#[repr(align(64))]
+struct MetaData {
```

Comment:
For unused field members please use the `unused` attribute, as we agreed in previous revisions.

Reply:
I don't recall such previous discussions.  Note this isn't a case where the compiler doesn't understand that a field is actually being used, this is a case where the field really isn't being used (yet, at least).

A precedent would be the scheduler's WakerPage struct, which has an unused field named "_unused" (it's there for padding) and no annotation.

Reply:
![image](https://user-images.githubusercontent.com/4939789/195677835-4ec6a853-5281-4f2a-8df2-3fb8c0d2b863.png)

Reply:
That case is different from this, as I explained in the comment.  In that case, the field is actually being used, just not in a way the compiler thought matters.  Using an underscore there would indicate that the field isn't being used, when it is.

This case, like the WakerPage case I cited, uses the underscore to indicate that a field isn't being used.  This is using the underscore in the way that it is intended to be used.